### PR TITLE
Word 365: do not announce font attribute toggle messages if raised from UIA notification event

### LINKED ISSUE #10950: Word 2019 last version duplicates NVDA attribute change notification when formatting text through keystrokes
Reproduced by many Spanish-speaking users, including me.

### Steps to reproduce:

1. Open a Microsoft Office Word document.
2. (Optional)Type any text.
3. Execute font attributes keystrokes (in Spanish, Control+N/S/K/etc.).

### Actual behavior:

NVDA starts announcing the change in the configured language but Word sends a notification with the attribute name in English and localized status interrupting NVDA, the same as happened with "delete back word" text.
IO - inputCore.InputManager.executeGesture (10:52:07.776):
Input: kb(laptop):control+n
IO - speech.speak (10:52:07.838):
Speaking [u'Negri\xf1a desactivada']
DEBUG - external:globalPlugins.wintenObjs.GlobalPlugin.uiaDebugLogging (10:52:07.871):
W10: UIA object: <NVDAObjects.Dynamic_WordDocumentEditableTextWithoutAutoSelectDetectionUIAWindowNVDAObject object at 0x054A0DD0>, name: , event: notification, app module: <'appModuleHandler' (appName u'winword', process ID 7392) at address 4d713b0>, automation Id: , class name: _WwG
DEBUG - external:globalPlugins.wintenObjs.GlobalPlugin.event_UIA_notification (10:52:07.871):
W10: UIA notification: sender: <POINTER(IUIAutomationElement) ptr=0x9f132d8 at 5486760>, notification kind: 2, notification processing: 3, display string: Bold Desactivado, activity ID: AccSN1
IO - speech.speak (10:52:07.878):
Speaking [u'Bold Desactivado']
Note: This happens with winTenApps disabled as well.

### Expected behavior:

NVDA ignores notification and announces changes by itself, or disables these announcements for the most recent Word 2019/365 releases and, despite it is not under our control, Word sends them in the configured language.

#### NVDA installed/portable/running from source:

Portable and installed

#### NVDA version:

2019.3.1

#### Windows version:

Windows 10 Version 1903 (build 18362.720).

#### Name and version of other software in use when reproducing the issue:

Microsoft Office Word 2019 Version 2003 (build 12624.20382), Office 365 License.

#### Does the issue still occur after restarting your computer?

Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

2019.2.1 - Reproducible

#### If addons are disabled, is your problem still occuring?

Yes

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?

N/A

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #10950 

### Summary of the issue:
In recent Word 365 releases, NVDA will announce font attribute toggle message up to two times - a message defined by NVDA ,and a second time thanks to UIA notification event.

### Description of how this pull request fixes the issue:
Builds on an earlier work on Word 365 UIA notification event suppression (#10851):

* Adds a dedicated list in UIA.WordDocument class listing activity ID's NVDA should ignore.
* Adds AccSN1, the activity Id for font attribute toggle messages.

### Testing performed:
Tested to make sure font attribute toggle messages are announced only once with Word 365, and verified that there was no regression with previous work on Word 365 fix.

### Known issues with pull request:
None

### Change log entry:
Nothing to add, but if a what's new entry should be provided, a wording similar to #10851 is ideal, this time talking about font attribute toggle commands such as toggling bold.

Thanks.

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved handling of Word 365 notifications to prevent unnecessary announcements for specific edit functions, enhancing the user experience by reducing noise.

### CODE REVIEW

File: source/NVDAObjects/UIA/wordDocument.py [hunk 1]
```diff
@@ -320,10 +320,15 @@ class WordDocument(UIADocumentWithTableNavigation,WordDocumentNode,WordDocumentB
 	# Microsoft Word duplicates the full title of the document on this control, which is redundant as it appears in the title of the app itself.
 	name=u""
 
+	noNotificationActivityIds = [
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/wordDocument.py:323]: would you mind renaming this to `supressedActivityIds`? If the term "Notification Activity" is relevant then add that to the comment for this list, something like "Notification Activity Ids that are suppressed to reduce unnecessary verbosity"

**Comment by michaelDCurran** [comment]
These notifications are needed in Windows 10 Mail and other future scenarios where the Office object model is unavailable. We should somehow still allow the notifications in these situations. 
The WinwordSelectionObject property on the NVDAObject for the document will be None in these scenarios.

**Comment by josephsl** [comment]
Hi, true. I think in the long term, we will need to think about what to do with font attribute toggle announcements for object model based implementation. This also means discussing translations with Microsoft. Thanks.

**Comment by josephsl** [comment]
Hi,

Regarding Windows 10 Mail, I can see what the issue is - I'm on it.

Thanks.

**Comment by feerrenrut** [comment]
Could this problem (differences in desired UIA notifications between mail and word) be addressed by fetching the list of ignored UIA notifications from the appmodule?

**Comment by josephsl** [comment]
Hi,

Possibly yes, but there are cases where it might not be the case:

* UIA objects may span across apps: different apps may use the same UIA object, especially if it is provided by XAML and other frameworks. If these elements raise UIA notification event, locating apps where this happens might become a challenge.
* Other ways to determine notification announcements: Word 365 comes with WinwordSelectionObject and does raise UIA notification event, whereas Windows 10 Mail does not come with this object nor fires notification events.

In the short term, it will be possible to let app modules tell NVDA whether or not UIA notification event should be passed onto UIA objects (app modules handle events before NVDA objects do). If it is implemented (I'm willing to do so once new feature PR freeze ends), it will lay the foundation to resolve Windows 10 Calculator bug (even then, Calculator app module must inform NVDA as to when to suppress UIA notification events). In the long term, app vendors should be reminded that UIA notification events are meant for essential information - Edge and Alarms and Clock apps are examples of good use of notification event.

Thanks for bringing this up.

**Comment by josephsl** [comment]
Hi,

2021 update: upon further testing, it appears a different approach is needed - a UIA notification event handler for Word app module, made more easier as there is already a proof of concept with a different app module (see Windows 10 App Essentials add-on). I'll rework this PR and perhaps target this for 2021.2 or later.

Thanks.

**Comment by josephsl** [comment]
Hi,

By the way, the rework plan also calls for transferring UIA notification event handler from Word document UIA object to Word app module unless UIA notification event is raised when changing fonts in other Office 365 applications.

Thanks.

**Comment by feerrenrut** [comment]
I'm going to mark this as a draft while work is on going.

**Comment by josephsl** [comment]
Hi, thanks – probably won’t work on this until 2021.2 cycle is under way.

**Comment by josephsl** [comment]
Hi,
Although I couldn't generate a debug log, it turns out UIA notification event can be seen in Outlook, therefore it justifies WordDocument object scope for dealing with UIA notification event.

Reopening the PR for review (targeting 2021.2).

**Comment by seanbudd** [comment]
>These notifications are needed in Windows 10 Mail and other future scenarios where the Office object model is unavailable. We should somehow still allow the notifications in these situations.
The WinwordSelectionObject property on the NVDAObject for the document will be None in these scenarios.

I'm hesitant to review this as I'm not confident this comment has been addressed. Is Windows 10 mail (and other scenarios Mick mentioned here that could arise) reporting this properly? Could we add system tests for Windows 10 mail?

**Comment by josephsl** [comment]
Hi, it will depend on what activity Id will be used by Windows 10 Mail for notifications like this. Also, you need to create an email account in order to test things with Mail, and it might not be available on the VM AppVeyor is using because Mail is a universal (modern) app with changes at least once a year (making the UI and UX a bit unpredictable; you can assume that smoke tests involving universal apps such as Windows 10 Mail app are a no go if AppVeyor VM uses Windows Server LTSC (2019 is the latest one at the moment) which do not come with Mail and other apps). Thanks.

**Comment by seanbudd** [comment]
@michaelDCurran - requesting your review here as per your previous comment, as we are unsure what the future impacts of merging this fix would be. 

> > These notifications are needed in Windows 10 Mail and other future scenarios where the Office object model is unavailable. We should somehow still allow the notifications in these situations.
> > The WinwordSelectionObject property on the NVDAObject for the document will be None in these scenarios.
> 
> I'm hesitant to review this as I'm not confident this comment has been addressed. Is Windows 10 mail (and other scenarios Mick mentioned here that could arise) reporting this properly?

**Comment by josephsl** [comment]
Hi, not that I know of. Like any universal app, Windows 10 Mail can change without notice (we had a debacle last year where Microsoft released a combined Mail and Calendar app with a different executable name, only to pull the experiment a few months later and the apps were separated again). As for UIA notification event, the most important thing is the activity Id, and the pull request I submitted would apply if Windows 10 Mail (current or future) raises UIA notification event with the same activity Id described. Thanks.

**Comment by cary-rowen** [comment]
Hi,
Can this PR fix #12101 ?
Thanks.

**Comment by josephsl** [comment]
Hi, I don’t think so, as this PR resolves extraneous announcements when keyboard commands are performed and UIA notification events are fired. Thanks.

**Comment by feerrenrut** [comment]
As discussed on #10283, there will need to be a resolution about preference of UIA notification vs detection via Object model. The preference may even need to vary depending on the version of windows or office.

**Comment by CyrilleB79** [comment]
> As discussed on #10283, there will need to be a resolution about preference of UIA notification vs detection via Object model. The preference may even need to vary depending on the version of windows or office.

In order to define in which version UIA should  be preferred:
* With Word 2016 version '16.0.5182.1000': No UIA notification at all
* With Word 2016 version '16.0.14430.20234' (up-to-date):
  * Common character formatting shortcuts (bold, italic, underline) are reported in localized version (French in my case)
  * However, if you put the caret between two spaces and press ctrl+I, you will hear "Italic disabled" in any case, whatever the state of the "Italic" button in the ribbon. This is a bug of this notifications.
  * Less common character formatting shortcuts are not reported (double underline, underline words only, all cap, smal caps, etc.)
  * Paragraph formatting shortcuts such as paragraph alignment or indenting are not reported.

So IMO, for an up-to-date Word 2016, we are not ready to switch to UIA.

I do not know if some more information is reported with Word 2019 or 365. But if you get the same version with those versions of Word, I would not switch to UIA today and recommend the following:
* Use object model when available
* Use UIA when object model not available
* In order to have a version where we could at least prefer UIA, report to Microsoft all the missing formatting information we want to have reported.

**Comment by josephsl** [comment]
Hi, build 14430 is actually Microsoft 365. Thanks.

**Comment by CyrilleB79** [comment]
> Hi, build 14430 is actually Microsoft 365. Thanks.

If I have understood correctly, Office 2016, 2019 and 365 share the same build numbers. That's why I have specified the Word version (2016) as well as the build number (14430) in my previous report.

@josephsl if you are able to test on Office 365 and if you have a little time, could you tell me if Office 365 offers more notifications than Office 2016 does?

**Comment by josephsl** [comment]
Hi, which commands would you like me to test? I’ll be able to provide technical info for these (I have Event Tracker add-on for this purpose). Thanks.

**Comment by CyrilleB79** [comment]
Hi Joseph
It would be interesting to test all the formatting script in , that is:
```
"kb:control+[":"increaseDecreaseFontSize",
"kb:control+]":"increaseDecreaseFontSize",
"kb:control+shift+,":"increaseDecreaseFontSize",
"kb:control+shift+.":"increaseDecreaseFontSize",
"kb:control+b":"toggleBold",
"kb:control+i":"toggleItalic",
"kb:control+u":"toggleUnderline",
"kb:control+=":"toggleSuperscriptSubscript",
"kb:control+shift+=":"toggleSuperscriptSubscript",
"kb:control+l":"toggleAlignment",
"kb:control+e":"toggleAlignment",
"kb:control+r":"toggleAlignment",
"kb:control+j":"toggleAlignment",
"kb:alt+shift+downArrow":"moveParagraphDown",
"kb:alt+shift+upArrow":"moveParagraphUp",
"kb:alt+shift+rightArrow":"increaseDecreaseOutlineLevel",
"kb:alt+shift+leftArrow":"increaseDecreaseOutlineLevel",
"kb:control+shift+n":"increaseDecreaseOutlineLevel",
"kb:control+alt+1":"increaseDecreaseOutlineLevel",
"kb:control+alt+2":"increaseDecreaseOutlineLevel",
"kb:control+alt+3":"increaseDecreaseOutlineLevel",
"kb:control+1":"changeLineSpacing",
"kb:control+2":"changeLineSpacing",
"kb:control+5":"changeLineSpacing",
["kb:control+m", "kb:control+shift+m", "kb:control+t", "kb:control+shift+t"]: script_changeParagraphLeftIndent
```

If it's too much, try to test a subset with the most common shortcuts.

Just tested on Word 2016, ('16.0.14430.20234') only ctrl+B, ctrl+I and ctrl+U report the formatting action via UIA. More specifically, I did not test directly checking the UIA event but checked that double speaking with speech history add-on.

Let me know for 365. Thanks.

**Comment by josephsl** [comment]
Hi, you’re right – only Control+B, Control+I, and Control+U causes Word to raise UIA notification evenT. Also, not related to this PR, but can we think of a way to convert these gestures entries to use script decorators in the future? Thanks.

**Comment by josephsl** [comment]
Hi, looks like my assumption was wrong in this case – Control+B/I/U does not report anything via UIA notification, evidenced by the fact that Narrator does not announce font changes when I press these keys in Outlook 365. Thanks.

**Comment by CyrilleB79** [comment]
> There is also a bug to note: If you press ctrl+B between two spaces, the UIA notification still says "Bold off" every time. This is not correct since:
> 
>     * the bold button state in the ribbon is toggled
> 
>     * bold is actually activated, i.e. if you start typing, the text will be bold
>       This bug should also be reported to Microsoft.
>       I have no contact with them. @josephsl could you take care of it?

I have reported to edad@microsoft.com and have received a reply from iSoftStone Inc (probably a contractor?) They write:
> Thank you so much for letting us know, we’ve informed the product team and a bug has been filed.  We will inform you of the outcome.

**Comment by josephsl** [comment]
Hi, the people I now from Microsoft are Windows engineers and folks working on accessibility (I’ll look around). Thanks.

**Comment by CyrilleB79** [comment]
I wrote earlier:
> To summarize, in newer versions of Office:
> 
>     1. UIA notifications are available in Word but not in Outlook
> 
>     2. Only ctrl+B, ctrl+I, ctrl+U are reported through UIA notifications, not all other formatting shortcuts
> 
>     3. Ctrl+B/I/U's result is not correctly reported when the cursor is located between two spaces.
> 
> 
> To start using UIA notifications in NVDA's codebase, I think that at least point 3. should be fixed by MS (or a workaround should be found) Regarding points 1. and 2., I think we'd better to wait for enhancements on these points in future versions. Implementing a version check and application check is not worth it today to have only ctrl+B/I/U reported in Word only (newer Office versions) and with the bug in point 3.
> 
> @feerrenrut given all these points, would you re-consider what you wrote in [#10990 (comment)](https://github.com/nvaccess/nvda/pull/10990#issuecomment-884715182) and prefer silencing UIA notifications for now, i.e. merging this PR as is? Or would you recommend something else? E.g. a configurable option in advanced settings?


April 2022 update:
Using Word 2016 version '16.0.15028.20204', the point 3 (bug providing erroneous toggle status when the caret is between two spaces) has been fixed. I do not know since when.

I do not know if Microsoft intend to fix points 1. and 2. and if they are aware of it.
@josephsl do you have any information about it? Or would you mind report it if not already?
Thanks.

**Comment by josephsl** [comment]
Hi, I have no info on it as I don’t track Office Insider builds. I think the PR is good to go unless NV Access (and Microsoft 365) folks says otherwise. Thanks.

**Review by seanbudd — DISMISSED** [review]

**Comment by lukaszgo1** [comment]
@seanbudd  Have you seen [this comment](https://github.com/nvaccess/nvda/pull/10990#issuecomment-613936474) by @michaelDCurran? As far as I can tell this still is not addressed and should be before this is merged.

**Comment by seanbudd** [comment]
> @seanbudd  Have you seen [this comment](https://github.com/nvaccess/nvda/pull/10990#issuecomment-613936474) by @michaelDCurran? As far as I can tell this still is not addressed and should be before this is merged.

@lukaszgo1 - this seems to contradict @josephsl's latest comment.
Reading through the thread it seems like this issue has been addressed.
It's unclear as to what is blocking this from being merged.
Could you provide an example, even if hypothetical, of what could go wrong by merging this?

**Comment by lukaszgo1** [comment]
@seanbudd  Merging the code as it is now would cause no announcement of bold / underline in cases where Office Object Model is not available i.e. when running MS Word in (Defender Application Guard). While the information reported via UIA is not as complete as what we can report with Object Model it is still better than no announcement at all. Ideally someone who has worked with Office apps running (in (Defender Application Guard)  perhaps @michaelDCurran  would be able to review this and decide if this can be merged.

**Comment by seanbudd** [comment]
Thanks @lukaszgo1, this will be reverted to a draft until the issue you described is fixed.

**Comment by josephsl** [comment]
Hi,

I can try on a machine (physical or virtual) with MDAG (Microsoft Defender Application Guard) to see what happens. At this stage, with Project Monarch on the horizon (coming soon), I feel it makes sense to proceed with this PR if MDAG compatibility can be verified.

Thanks (p.s. I'm back from vacation, although I won't do significant coding until July).

**Comment by XLTechie** [comment]
> 3. Ctrl+B/I/U's result is not correctly reported when the cursor is located between two spaces.

How can a cursor be between two spaces? It is either on one space, or it is on 
the other, never between.

Do you mean on the second space of a two space set? I.e. on the third character 
of "a  b"

**Comment by CyrilleB79** [comment]
> How can a cursor be between two spaces? It is either on one space, or it is on the other, never between. Do you mean on the second space of a two space set? I.e. on the third character of "`a  b`"

I meant visually.
In Word as in any edit field, the cursor, a.k.a. "insertion point" is visually a blinking vertical bar positioned where the text will be inserted. So in your example "`a  b`", it is positioned after the first space and before the second one.
In braille anywhere or visually in Windows console, the cursor is positioned "on" (or "below") a character and you know that the text you type is inserted before the current character (if insert mode is enabled which is the far most frequent usage). In your example, the braille cursor is actually on the second space of a two space set? I.e. on the third character of "`a  b`" as you wrote.

I hope to have clarified things.

**Comment by josephsl** [comment]
Hi,

Ping @MichaelDCurran on the status of MDAG + Word 365, along with other participatns on the status of this PR.

Thanks.

**Comment by seanbudd** [comment]
@josephsl - apologies we missed this. I'm marking this as ready for review.
Can you give us an update and ask any pending questions you have from us?

**Comment by josephsl** [comment]
Hi,

Any ideas on MDAG compatibility? If not, I think we could proceed - may need to update the branch to latest master before merging it.

Thanks.

**Comment by seanbudd** [comment]
I don't think we have any special knowledge here, I'd encourage getting manual testing to confirm

**Comment by josephsl** [comment]
Hi, I don’t have access to MDAG+Word 365 combination at all (wish I did). Thanks.

**Comment by seanbudd** [comment]
I think this PR needs wider testing then. Can anyone test these scenarios and report back?

@coderabbit can you summarise the discussion on this PR?

**Comment by seanbudd** [comment]
I think this PR needs wider testing then. Can anyone test these scenarios and report back?

@coderabbitai can you summarise the discussion on this PR?

**Comment by Adriani90** [comment]
I cannot reproduce these mistranslated examples reported in #10950 with NVDA last alpha and Microsoft 365 MSO (Version 2405 Build 16.0.17628.20006) 64 Bit in german or romanian language.
But those messages are probably properly translated by Microsoft in these languages. It seems only spanish was impacted in that issue.

@ivnc, @nvdaes could you test whether the issue with double reportings is still occuring for you in the wrong translation?
If the translation is now correct, I think this PR is not really needed anymore.
The message provided by NVDA is interupted by the UIA notification, but it is really hard to distinduish that there are two messages spoken actually unless you look into the log file.
I think instead of dropping the whole UIA notifications, it might be better to contact Microsoft and suggest to them to translate the UIA messages in your Office language properly.

**Comment by nvdaes** [comment]
Now I can listen right translated messages, but I distinguish that the start of the message is briefly interrupted, repeating the first character. I've tested control+n (bold/negrita), control+s (underline/subrayado), and control+k (italic/cursiva), in Word 365, Windows 11.

**Comment by Adriani90** [comment]
In this case it might be better to rather stick to the UIA notifications and make NVDA supress its own messages in these cases for Office 365 in case the current behavior is disturbing.Von meinem iPhone gesendetAm 31.05.2024 um 07:14 schrieb Noelia Ruiz Martínez ***@***.***>:﻿
Now I can listen right translated messages, but I distinguish that the start of the message is briefly interrupted, repeating the first character. I've tested control+n (bold/negrita), control+s (underline/subrayado), and control+k (italic/cursiva), in Word 365, Windows 11.

**Comment by josephsl** [comment]
Hi, for consumers, that solution might work since people might be using newer Microsoft 365 releases. However, don’t forget about our recent discussion of corporate mode where we discussed tendency for enterprises to use older software releases – some of them using older Word versions without UIA notification event handler included. Thanks.

**Comment by Adriani90** [comment]
Is it possible to supress NVVDA messages for MS Office 365 only?Von meinem iPhone gesendetAm 31.05.2024 um 15:22 schrieb Joseph Lee ***@***.***>:﻿
Hi, for consumers, that solution might work since people might be using newer Microsoft 365 releases. However, don’t forget about our recent discussion of corporate mode where we discussed tendency for enterprises to use older software releases – some of them using older Word versions without UIA notification event handler included. Thanks.

**Comment by CyrilleB79** [comment]
@Adriani90 the situation is more complex. The decision to use object model or UIA notifications should be base on various parameters: Office version (2013 / 2016 / 365, etc.), Office build version, software used (Word or Outlook), Windows version (8.1 or 10/11) and running in MDAG or not.

You need to read all the comments for more context; be careful: given the high number of comments, some of them are hidden so you'll have to display them.

**Comment by Adriani90** [comment]
In my case I tested with objectModel and still the UIA notifications are translated properly. MDAG I will test as well soon, but you can be sure that in 365 the UIA notifications are fired properly with objectModel and also when UIA is enabled. You can test yourself.In Office 2016, the NVDA messages should definitely remain.Von meinem iPhone gesendetAm 31.05.2024 um 17:48 schrieb Cyrille Bougot ***@***.***>:﻿
@Adriani90 the situation is more complex. The decision to use object model or UIA notifications should be base on various parameters: Office version (2013 / 2016 / 365, etc.), Office build version, software used (Word or Outlook), Windows version (8.1 or 10/11) and running in MDAG or not.
You need to read all the comments for more context; be careful: given the high number of comments, some of them are hidden so you'll have to display them.

**Comment by CyrilleB79** [comment]
> In my case I tested with objectModel and still the UIA notifications are translated properly. MDAG I will test as well soon, but you can be sure that in 365 the UIA notifications are fired properly with objectModel and also when UIA is enabled. You can test yourself.In Office 2016, the NVDA messages should definitely remain.

@Adriani90 to be clear and avoid any confusion, UIA vs object model notifications upon formatting command are by no mean controlled by the option "Use UI Automation to access Microsoft Word document controls" in advanced settings.

**Comment by seanbudd** [comment]
We're considering trialling this PR for wider testing by merging early in an alpha cycle. Any thoughts @josephsl ?

**Comment by josephsl** [comment]
Hi, let’s get the show going – feel free to merge it whenever you are ready (while I may not respond to all issues arising from it, I’ll keep an eye on it from time to time; do ping me if I do need to resolve some things). Thanks.

**Pull request merged**